src/docfx/config/Config.cs

```diff
@@ -193,68 +197,89 @@ private static void ValidateContributorConfig(ContributionConfig config, string
             {
-                var extendConfigPath = PathUtility.NormalizeFile(Path.Combine(Path.GetDirectoryName(configPath), path));
-                extendedConfig.Merge(LoadOriginalConfigObject(extendConfigPath, parents, false));
+                return config;
             }
-            extendedConfig.Merge(config);
-            parents.RemoveAt(parents.Count - 1);
-
-            return extendedConfig;
-        }
 
-        private static IEnumerable<string> GetExtendConfigPaths(JToken objExtend)
-        {
-            if (objExtend == null)
-                yield break;
-            if (objExtend is JValue strExtend)
+            if (!(objExtend is JArray arrayExtend))
             {
-                yield return strExtend.Value.ToString();
-                yield break;
+                return config;
             }
-            if (objExtend is JArray arrExtend)
+
+            var extendedConfig = new JObject();
+            foreach (var extendPath in arrayExtend)
             {
-                foreach (var path in arrExtend)
+                if (extendPath is JValue strExtendPath)
                 {
-                    if (!(path is JValue strPath))
-                        throw new Exception($"Expect to be string: {JsonUtility.Serialize(path)}");
-                    yield return strPath.Value.ToString();
+                    var filePath = restoreMap.GetUrlRestorePath(docsetPath, strExtendPath.Value<string>());
+                    var configObject = LoadConfigObject(docsetPath, filePath, false, restoreMap); // only support first level extends
+                    extendedConfig = JsonUtility.Merge(extendedConfig, configObject);
                 }
-                yield break;
             }
-            throw new Exception($"Expect 'extend' to be string or array: {JsonUtility.Serialize(objExtend)}");
+
+            return JsonUtility.Merge(extendedConfig, config);
         }
 
         private static JObject ExpandAndNormalize(JObject config)
         {
             config[ConfigConstants.Content] = ExpandFiles(config[ConfigConstants.Content]);
             config[ConfigConstants.FileMetadata] = ExpandGlobConfigs(config[ConfigConstants.FileMetadata]);
             config[ConfigConstants.Routes] = ExpandRouteConfigs(config[ConfigConstants.Routes]);
+            config[ConfigConstants.Extend] = ExpandExtend(config[ConfigConstants.Extend]);
             config[ConfigConstants.Redirections] = NormalizeRedirections(config[ConfigConstants.Redirections]);
             config[ConfigConstants.RedirectionsWithoutDocumentId] = NormalizeRedirections(config[ConfigConstants.RedirectionsWithoutDocumentId]);
             return config;
         }
 
+        private static JToken ExpandExtend(JToken extend)
+        {
+            if (extend == null)
+                return null;
+
+            if (extend is JValue extendValue)
+            {
+                extend = new JArray(extendValue);
+            }
+
```

Comment:
can use `ExpandStringArray`

Reply:
no error handling in that method but it still can be used. ;)

---
In reply to: [203667192](https://github.com/dotnet/docfx/pull/3101#discussion_r203667192) [](ancestors = 203667192)